### PR TITLE
QuerySets: filter permissions by organizations

### PR DESCRIPTION
**Pull request by stsewd**
With the aim to bring logic from .com to .org more closer
I'm moving the organization's filters to .org.
They are going to be used if `RTD_ALLOW_ORGANIZATIONS` is True,
but they eventually will be merged with the current querysets
when organizations and normal projects are supported in .org.

In .com we are relying on user.projects having the
the projects where the user is member of and organization owner
(we are syncing this with signals).
We don't rely on that hack anymore and always check from the
organization models. The signals won't be removed for now,
but shouldn't be needed anymore.

To re-use more code, I have brought the SSO concept here,
but isn't implemented yet, we can bring the SSO models and logic later
easily.

This could be seen as more queries,
but the main ones were already being executed in .com (Project/VersionQueryset).
The other ones are executed only when using the API V2,
so I don't think that would be a problem.

I started this refactor using a mixin,
but then we would need to override every single
queryset in .com, using composition only requires
one override (already used),
and doesn't bring other methods into the class.

### CODE REVIEW

File: readthedocs/builds/querysets.py [hunk 1]
```diff
@@ -27,16 +28,16 @@ class VersionQuerySetBase(models.QuerySet):
...
-
-        .. note::
-
-           In .org all users are admin and member of a project.
-           This will change with organizations soon.
-        """
-        if user.is_authenticated:
-            projects_pk = user.projects.all().values_list('pk', flat=True)
+        """Add related objects from projects where `user` is an `admin` or a `member`."""
+        if user and user.is_authenticated:
```
Comment (stsewd) [inline readthedocs/builds/querysets.py:32]: had to add the user check everywhere 'cause we are passing user as None in several parts from .com :/

File: readthedocs/core/permissions.py [hunk 2]
```diff
@@ -1,18 +1,86 @@
...
 
 from readthedocs.core.utils.extend import SettingsOverrideObject
+from readthedocs.organizations.constants import ADMIN_ACCESS, READ_ONLY_ACCESS
 
 
 class AdminPermissionBase:
 
     @classmethod
-    def admins(cls, obj):
+    def projects(cls, user, admin=False, member=False):
```
Comment (stsewd) [inline readthedocs/core/permissions.py:13]: All this was extracted from .com

File: readthedocs/settings/base.py [hunk 3]
```diff
@@ -610,6 +610,7 @@ def DOCKER_LIMITS(self):
     DEFAULT_VERSION_PRIVACY_LEVEL = 'public'
     GROK_API_HOST = 'https://api.grokthedocs.com'
     ALLOW_ADMIN = True
+    RTD_ALLOW_ORGANIZATIONS = False
```
Comment (stsewd) [inline readthedocs/settings/base.py:613]: This setting will change to `RTD_ALLOW_ONLY_ORGANIZATONS` at some point when we support both single projects and organizations in .org :)

File: readthedocs/core/permissions.py [hunk 4]
```diff
@@ -1,18 +1,86 @@
...
+        return projects
+
+    @classmethod
+    def _has_sso_enabled(cls, organization):
+        return False
+
+    @classmethod
+    def _get_projects_for_sso_user(cls, user, admin=False):
+        from readthedocs.projects.models import Project
+        return Project.objects.none()
```
Comment (stsewd) [inline readthedocs/core/permissions.py:78]: don't really like having these as placeholders, but was this or repeat the whole thing in .com or block this till we move the vcs integration.

File: readthedocs/core/permissions.py [hunk 5]
```diff
@@ -1,18 +1,86 @@
...
+        if not settings.RTD_ALLOW_ORGANIZATIONS:
+            # All users are admin and member of a project
+            # when we aren't using organizations.
+            return user.projects.all()
+
+        if admin:
+            # Project Team Admin
+            admin_teams = user.teams.filter(access=ADMIN_ACCESS)
+            for team in admin_teams:
+                if not cls._has_sso_enabled(team.organization):
```
Comment (humitos) [inline readthedocs/core/permissions.py:48]: It seems that you removed the `provider=` attribute from here. Why? (also in the other calls to `has_sso_enabled` from this method)

File: readthedocs/core/permissions.py [hunk 6]
```diff
@@ -1,18 +1,86 @@
...
+            for team in member_teams:
+                if not cls._has_sso_enabled(team.organization):
+                    projects |= team.projects.all()
+
+            projects |= cls._get_projects_for_sso_user(user, admin=False)
+
+        return projects
+
+    @classmethod
+    def _has_sso_enabled(cls, organization):
```
Comment (humitos) [inline readthedocs/core/permissions.py:72]: This method should be public (without the `_` prefix), since it's used in different places outside these querysets.

Comment (humitos) [reply readthedocs/core/permissions.py:72]: Also, it should receive `provider=` to check if has VCS or Google SSO

File: readthedocs/core/permissions.py (hunk 4 above)
Comment (humitos) [reply readthedocs/core/permissions.py:78]: Why don't bring `CorporateAdminPermission._get_projects_for_sso_user` here and use the setting that you just added `RTD_ALLOW_ORGANIZATIONS` to decide what to do?

Comment (humitos) [reply readthedocs/core/permissions.py:78]: In any case, if we are already planning to change the content of this method, I'd like to see a `TODO:` list explaining the plan here and what's the code we need to bring here.

**Review by humitos — CHANGES_REQUESTED** [review]
I like this change. Using more `AdminPermission` is good. Hopefully, we end having all the permission login in one place someday 😄 

I'd like to see the changes I'm asking for in the comments before merging this. It will be hard to re-review this code again once we have SSO working on .org. I prefer to call the methods/functions as we expect to be called in the future instead of having to come back here to add an attribute (e.g. `provider=` in `has_sso_enabled`)

Finally, seems weird to read "This code was copied from ..." but then find there are changes in the code. It doesn't feel trustable.

File: readthedocs/core/permissions.py (hunk 2 above)
Comment (stsewd) [reply readthedocs/core/permissions.py:13]: yep

File: readthedocs/core/permissions.py (hunk 4 above)
Comment (stsewd) [reply readthedocs/core/permissions.py:78]: >  Why don't bring CorporateAdminPermission._get_projects_for_sso_user here and use the setting that you just added RTD_ALLOW_ORGANIZATIONS to decide what to do?

It has references to models that we don't have here.

> I'd like to see a TODO: list explaining the plan here and what's the code we need to bring here.

This is already overriden in .com.

File: readthedocs/core/permissions.py (hunk 5 above)
Comment (stsewd) [reply readthedocs/core/permissions.py:48]: We don't have the model that has those constants here, it was always using the email provider to check things, that is passed from .com.

File: readthedocs/core/permissions.py (hunk 6 above)
Comment (stsewd) [reply readthedocs/core/permissions.py:72]: This is only used here, it acts as a proxy for the `has_sso_enabled` function from .com, is not the same.

Comment (humitos) [reply readthedocs/core/permissions.py:73]: I want to reduce the overrides for the same method/function name. With this change, we have 3 `has_sso_enabled` that do different things. There are 2 that call the "original" one with different arguments; which is confusing and hard to follow. I prefer if we don't add these 2 extras and always use the same one.

Comment (stsewd) [reply readthedocs/core/permissions.py:73]: what about making provider optional `None` and default to ALL_AUTH? but looks like calling has_sso_enabled with None is valid... so what about naming this `has_allauth_sso_enabled`?

Comment (humitos) [reply readthedocs/core/permissions.py:73]: I'd say no to this proposal. These are my reasons:

- calling with `None` checks for "any type of SSO enabled"
- making the call explicit where it's used is better to understand what's happening in that code and easier to read
- we currently have only two (ALL_AUTH and GOOGLE), but this may change in the near future
  - SAML
  - OKTA
  - split ALL_AUTH into GITHUB, GITLAB, BITBUCKET; since I found cases where this will be useful

File: readthedocs/core/permissions.py [hunk 7]
```diff
@@ -4,10 +4,82 @@
...
+        return projects
+
+    @classmethod
+    def has_sso_enabled(cls, obj, provider=None):
+        return False
+
+    @classmethod
+    def _get_projects_for_sso_user(cls, user, admin=False):
+        from readthedocs.projects.models import Project
+        return Project.objects.none()
```
Comment (stsewd) [inline readthedocs/core/permissions.py:81]: These two depend on the SSO implementation, they are overridden on .com.

File: readthedocs/core/permissions.py (hunk 4 above)
Comment (humitos) [reply readthedocs/core/permissions.py]: Now that we have the models in community, we can move the code from corporate here and use the `RTD_ALLOW_ORGANIZATIONS` as suggested, right? That way we will have all the code in the same place and once we enable organization in community it will just work.

**Comment by humitos** [comment]
This is looking good. I ask a question at https://github.com/readthedocs/readthedocs.org/pull/8298#discussion_r691090513 because I think it makes sense to bring that corporate code here at this point before merging.

**Comment by stsewd** [comment]
@humitos didn't move those because it includes very specific code related to the implementation, and it also includes checks for google, which we won't have in .org I think.

**Comment by ericholscher** [comment]
Final review waiting on @humitos 👍

**Review by humitos — APPROVED** [review]
I'm fine merging this, but I still think is worth removing the override from corporate on those methods as I mentioned in my previous comment.

These are the methods that are overwritten from corporate: https://github.com/readthedocs/readthedocs-corporate/pull/1232/files#diff-d040f1c8c0a21fef1c927a0070abb1f47b46df75075141fba8af6bd0acd7feb0R37

and I think we can add an `if settings.RTD_ALLOW_ORGANIZATIONS` in community to execute one or the other. That way, we are removing an override and reducing the complexity while sharing more code between instances. I don't clearly see why we can't do this. Besides, once we enable GH SSO on community, it will just work.

Google provider check won't be a problem since there won't be organizations with it in community, so I will be just skipped.
